### PR TITLE
Add sandbox API for task insertion to service LB and service discovery

### PR DESCRIPTION
**Pull request by sanimej**
This change [was added earlier](https://github.com/docker/libnetwork/pull/1416) but reverted since the docker changes weren't ready.

In this version there are some minor changes to address the comments raised in the docker/docker review, https://github.com/docker/docker/pull/27279

Signed-off-by: Santhosh Manohar santhosh@docker.com

### CODE REVIEW

File: networkdb/networkdb.go [hunk 1]
```diff
@@ -227,10 +227,14 @@ func (nDB *NetworkDB) getEntry(tname, nid, key string) (*entry, error) {
 // table, key) tuple and if the NetworkDB is part of the cluster
 // propogates this event to the cluster. It is an error to create an
 // entry for the same tuple for which there is already an existing
-// entry.
+// entry unless the current entry is deleting state.
 func (nDB *NetworkDB) CreateEntry(tname, nid, key string, value []byte) error {
-	if _, err := nDB.GetEntry(tname, nid, key); err == nil {
-		return fmt.Errorf("cannot create entry as the entry in table %s with network id %s and key %s already exists", tname, nid, key)
+	oldEntry, err := nDB.getEntry(tname, nid, key)
+	if err != nil || oldEntry != nil && !oldEntry.deleting {
```
Comment (aboch) [inline networkdb/networkdb.go]: I feel the following would be clearer to read:
```
if A {
      return err1
}
if  B {
    return err2
}
```

File: endpoint.go [hunk 2]
```diff
@@ -303,6 +304,18 @@ func (ep *endpoint) isAnonymous() bool {
 	return ep.anonymous
 }
 
+// enableSerice sets ep's serviceEnabled to the passed value if its not in the
```
Comment (dongluochen) [inline endpoint.go]: nit: `enableSerice` -> `enableService`
`if its not` -> `if it's not`

**Comment by dongluochen** [comment]
SGTM (not a maintainer)

File: sandbox.go [hunk 3]
```diff
@@ -655,6 +661,30 @@ func (sb *sandbox) SetKey(basePath string) error {
 	return nil
 }
 
+func (sb *sandbox) EnableService() error {
+	for _, ep := range sb.getConnectedEndpoints() {
+		if ep.enableService(true) {
+			if e := ep.addToCluster(); e != nil {
```
Comment (stevvooe) [inline sandbox.go]: Can we use the appropriate variable name for errors here? `err`.

File: endpoint.go (hunk 2 above)
Comment (sanimej) [inline endpoint.go]: @dongluochen fixed the typo.

File: sandbox.go (hunk 3 above)
Comment (sanimej) [inline sandbox.go]: @stevvooe changed the variable name to `err`

**Comment by thaJeztah** [comment]
ping @aboch @stevvooe PTAL, this is blocking a PR on the 1.13 milestone

**Comment by aboch** [comment]
LGTM

**Pull request merged**